Mathlib/RingTheory/Spectrum/Prime/ChevalleyComplexity.lean

```diff
@@ -0,0 +1,879 @@
+/-
+Copyright (c) 2025 Yaël Dillies, Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Andrew Yang
+-/
+import Mathlib.Algebra.Order.SuccPred.WithBot
+import Mathlib.Algebra.Polynomial.CoeffMem
+import Mathlib.Data.DFinsupp.WellFounded
+import Mathlib.RingTheory.Spectrum.Prime.ConstructibleSet
+import Mathlib.RingTheory.Spectrum.Prime.Polynomial
+
+/-!
+# Chevalley's theorem with complexity bound
+
+Chevalley's theorem states that if `f : R → S` is a finitely presented ring hom between commutative
+rings, then the image of a constructible set in `Spec S` is a constructible set in `Spec R`.
+
+Constructible sets in the prime spectrum are made of open sets in the prime spectrum, which are
+themselves made from a family of polynomials.
+
+We say an open set *has complexity at most `M`* if it can be written as the zero locus of a family
```

Comment:
Would it be worthy to try to track degree and number of open sets separately?

Reply:
Do you mean "number of polynomials" rather than "number of open sets"? We're tracking the former, not the latter.

Given the shape of the bound (a tower of size depending on both the degree and number of polynomials used to describe the original variety), I doubt it is possible to track both variables independently.